### ISSUE TITLE
Retire preallocation model and memory settings

### ISSUE THREAD

**Issue by erkyrath** [DavidKinder/Inform6#78]
As long as we're recording long-term goals... I always meant to replace all the preallocation limits (ALLOC_CHUNK_SIZE, etc) with realloc() calls. 

I may never do this, but at least it's in the tracker now.

Related: There are many places where a bunch of parallel allocated arrays could be replaced by a single realloced array of structs. For example, symbs/svals/sflags/stypes/slines/smarks are all allocated with length MAX_SYMBOLS. Similarly, label_offsets/label_symbols/label_next/label_prev are all allocated with length MAX_LABELS. Structs would make this all easier to deal with.

**Comment by Kroc** [comment]
It's doubtful it would be a problem, but how would this affect low-memory systems? I'm running Inform6 on a Psion 5, which is an 18MHz ARM3 with roughly 1-4 MB of RAM readily usable (the RAM is shared with the RAM-disk). The test-suite provided with PunyInform is not able to compile due to size, but that's not a concern, I don't think I could write any game that could reach those limits.

**Comment by erkyrath** [comment]
It would probably increase the memory requirements by some fraction. Allocations could start lower, but the jumps would be coarser and you'd need some headroom for the realloc copying. Overall, slightly more expensive.

**Comment by erkyrath** [comment]
Also related: MAX_LINES_PER_VERB (32) is currently hardwired into the code. It would be easy to make this auto-resize as well. (It's the length of an array in `struct verbt`, not a global array.)

**Comment by erkyrath** [comment]
Progress on this: the following settings have been removed as of right now.

ALLOC_CHUNK_SIZE
SYMBOLS_CHUNK_SIZE
MAX_ABBREVS (removed only for Glulx)
MAX_CLASSES
MAX_OBJECTS
MAX_SYMBOLS

More to come.

**Comment by erkyrath** [comment]
So what's left? Only the hard cases!

Grepping for memoryerror() turns up the following constants: MAX_GLOBAL_VARIABLES, MAX_QTEXT_SIZE, MAX_STATIC_STRINGS, MAX_LOW_STRINGS, MAX_DYNAMIC_STRINGS.

---

MAX_DYNAMIC_STRINGS is a tradeoff against MAX_ABBREVS in Z-code. So those settings both remain.

In Glulx, MAX_DYNAMIC_STRINGS is limited to 100 (because the string escapes for dynamic strings are `"@00"` through `"@99"`). I will set the default value for MAX_DYNAMIC_STRINGS to 100 in Glulx; it will still be a settable value but nobody will need to change it.

---

MAX_GLOBAL_VARIABLES is fixed in Z-code. In Glulx, we have an annoying situation where the compiler reserves space for MAX_GLOBAL_VARIABLES globals even if they're not all used. If I make this dynamic, it changes the build output for existing games, which gives me a headache.

I think what I want to do is make it dynamic, but with a minimum of 512. (The current default.) That way existing games will compile the same, but a game with more globals will still compile. This doesn't allow people to scrimp bytes, but Glulx is not for byte-scrimpers.

---

MAX_QTEXT_SIZE is a *real* headache. This is used to allocate `lexeme_memory`, a rolling circular buffer used to parse tokens. All `token_data` structs have a `char *text` pointer that points somewhere into `lexeme_memory` -- unless it points somewhere else. There's no notion of "owning" this text memory. Pointers to it are juggled and copied freely.

This is a nightmare and I have no intention of getting involved.

(I have a pending change to pass `const token_data *` pointers to various utility functions, instead of passing complete `token_data` structs. This will at least simplify the world for future maintainers.)

---

MAX_LOW_STRINGS and MAX_STATIC_STRINGS are a different headache. These are used for `low_strings` and `strings_holding_area`, which are passed to translate_text(). Actually `zcode_holding_area` is passed to translate_text() too.

The problem is that translate_text() takes a fixed-size buffer; it has no notion of resizing the buffer that you pass it. Sometimes its argument really *is* fixed-size. Sometimes it's one of the buffers named above, which we *could* resize, but translate_text() doesn't know how to do it.

Enhancing translate_text() seems possible, but it would be pretty messy. I am undecided about the idea.

**Comment by DavidKinder** [comment]
Once MAX_LOW_STRINGS is gone, we can also remove the command line arguments for memory size ($huge, $large and $small) and related logic, since they will no longer have any effect.

**Comment by erkyrath** [comment]
Hey, lookitthat.

Really we could do that *now* if we edged the $SMALL default for MAX_LOW_STRINGS up to 2048. Which would be a pretty minor change.

(I'm still looking at the translate_text() problem.)

**Comment by erkyrath** [comment]
translate_text() turned out to be pretty simple. Now we're just left with MAX_QTEXT_SIZE.

I wasn't kidding about this being a nightmare. As tokens are parsed, they are stored in a circular buffer which defaults to 20k bytes. This is guaranteed to have room for five tokens, by setting a maximum of 4k for quoted strings. But the average case for tokens is one or two bytes (operators) or up to 32 (identifiers) so it holds more like 3000-5000 tokens in normal use.

As the comment in lexer.c notes:

> ...pending some better method of garbage collection, we simply use a buffer so large that unless expressions spread across 10K of source code are found, there can be no problem.

Now that I understand the code, I'm sure I could construct examples which *are* a problem. Give me a couple of days to work on it.

I have now made two attempts to replace the circular buffer with an extendable data structure. I have failed twice. I am really sucking the mop here.

The problem is the "garbage collection" alluded to above. Once a token is generated, the compiler feels free to use that text pointer all over the place. It gets copied into global and local variables, stored in tables... and its lifespan is "Eh, probably long enough." This rots my socks.

The most conservative strategy is to store these token texts forever. This works but it's *extremely* slow and hogs memory out the wazoo.

I figured the conservative-and-practical strategy is to release all the token texts after each top-level directive and function is complete. This *almost* works, but large (I7-generated) source files build wrong. (Also, it's still somewhat slower than the original circular buffer. I haven't done measurements yet -- no point while it's still buggy.)

My next step is to audit all uses of `token_text` and look for places where it gets stashed for an excessively long time. Perhaps it's practical to clean those up (by copying to a more permanent buffer). Perhaps not.

A full-on reference counting strategy would work, but it would require fixing up *every* use of `token_text`. I fear this plan. Subtle bugs would sneak in.

If I can't find a plan I'm confident in, MAX_QTEXT_SIZE may have to stay as-is. (I will still PR some code cleanups, particularly if I succeeded in creating sme examples that break the current compiler.)

**Comment by erkyrath** [comment]
I feel I have to be fair by quoting Graham's commentary from the Tech Manual:

> Unfortunately code generation time may not be for a while, and this raises a problem: we clearly need to keep lexeme text in memory for a while, but it would be very costly on memory to keep the text of every token in memory permanently.  When is it safe for the lexer to throw a lexeme away?
> One solution would be for a formal system by which the code generator explicitly deallocated the space, but this is too bureaucratic and too risky (suppose we miss 0.1% of these? Memory will slowly clog up and begin to cause odd errors on large compilation runs).  Instead, the lexer simply writes its text cyclically into a large buffer.  Code generation almost always takes place before the lexer has got any more than 100 or so bytes of text ahead; since the buffer is a good 10K long, there is a generous safety margin.

For all my grumping, this is a sensible compromise. I agree entirely (see last comment) that the formal tracking plan is "too bureaucratic and too risky".

**Comment by erkyrath** [comment]
Turned out that local variables (and the stackargs bug, c.f. https://github.com/DavidKinder/Inform6/issues/122) was the big stumbling block here. Once I cleaned that up, my token-cleaning approach started working correctly. (In fact, it's sufficient to clean out token texts after each *statement*. So the lextexts list never gets very long at all.)

PR to clean up MAX_QTEXT_SIZE: https://github.com/DavidKinder/Inform6/pull/125

With this in place, the memory limits are history. I'm going to do one more PR after this to remove the memoryerror() call entirely. (It's still used for MAX_DYNAMIC_STRINGS, but we can generate a more specific error for that case.)
